include/reduce_helper.h

```diff
@@ -0,0 +1,237 @@
+using count_t = unsigned int;
+#endif
+
+namespace quda
+{
+
+  namespace reducer
+  {
+    /** returns the reduce buffer size allocated */
+    size_t buffer_size();
+
+    void *get_device_buffer();
+    void *get_mapped_buffer();
+    void *get_host_buffer();
+    count_t *get_count();
+    cudaEvent_t &get_event();
+  } // namespace reducer
+
+  /**
+     @brief The initialization value we used to check for completion
+   */
+  template <typename T> constexpr T init_value() { return -std::numeric_limits<T>::infinity(); }
+  template <typename T> struct atomic_type {
+    using type = device_reduce_t;
+  };
+  template <> struct atomic_type<float> {
+    using type = float;
+  };
+
+  template <typename T> struct ReduceArg {
+    const int n_reduce; // number of reductions of length n_item
+#ifdef HETEROGENEOUS_ATOMIC
+    using system_atomic_t = typename atomic_type<T>::type;
+    static constexpr int n_item = sizeof(T) / sizeof(system_atomic_t);
+    cuda::atomic<T, cuda::thread_scope_device> *partial;
+    // for heterogeneous atomics we need to use lock-free atomics -> operate on scalars
+    cuda::atomic<system_atomic_t, cuda::thread_scope_system> *result_d;
+    cuda::atomic<system_atomic_t, cuda::thread_scope_system> *result_h;
+#else
+    T *partial;
+    T *result_d;
+    T *result_h;
+#endif
+    count_t *count;
+
+    ReduceArg(int n_reduce = 1) :
+      n_reduce(n_reduce),
+      partial(static_cast<decltype(partial)>(reducer::get_device_buffer())),
+      result_d(static_cast<decltype(result_d)>(reducer::get_mapped_buffer())),
+      result_h(static_cast<decltype(result_h)>(reducer::get_host_buffer())),
+      count {reducer::get_count()}
+    {
+      // check reduction buffers are large enough if requested
+      auto max_reduce_blocks = 2 * deviceProp.multiProcessorCount;
+      auto reduce_size = max_reduce_blocks * n_reduce * sizeof(*partial);
+      if (reduce_size > reducer::buffer_size())
+        errorQuda("Requested reduction requires a larger buffer %lu than allocated %lu", reduce_size,
+                  reducer::buffer_size());
+
+#ifdef HETEROGENEOUS_ATOMIC
+      if (!commAsyncReduction()) {
+        // initialize the result buffer so we can test for completion
+        for (int i = 0; i < n_reduce * n_item; i++) {
+          new (result_h + i) cuda::atomic<system_atomic_t, cuda::thread_scope_system> {init_value<system_atomic_t>()};
+        }
+        std::atomic_thread_fence(std::memory_order_release);
+      } else {
+        // write reduction to GPU memory if asynchronous (reuse partial)
+        result_d = nullptr;
+      }
+#else
+      if (commAsyncReduction()) result_d = partial;
+#endif
+    }
+
+    template <typename host_t, typename device_t = host_t> void complete(host_t *result, const qudaStream_t stream = 0)
+    {
+#ifdef HETEROGENEOUS_ATOMIC
+      for (int i = 0; i < n_reduce * n_item; i++) {
+        while (result_h[i].load(cuda::std::memory_order_relaxed) == init_value<system_atomic_t>()) {}
```

Comment:
could we use `result_h[i].wait(init_value<system_atomic_t>(),cuda::std::memory_order_relaxed)` here ?

Reply:
I hadn't realized this was included in CUDA 11.0's libcu++ since it's a C++20 feature, but I see it's there.  Will fix this in next push.